components/engine/engine-camel/src/main/java/org/eclipse/dirigible/components/engine/camel/invoke/Invoker.java

```diff
@@ -11,37 +11,68 @@
 package org.eclipse.dirigible.components.engine.camel.invoke;
 
 import org.apache.camel.Message;
+import org.apache.camel.component.platform.http.springboot.PlatformHttpMessage;
 import org.eclipse.dirigible.components.engine.camel.processor.CamelProcessor;
 import org.eclipse.dirigible.graalium.core.DirigibleJavascriptCodeRunner;
 import org.eclipse.dirigible.graalium.core.javascript.CalledFromJS;
+import org.graalvm.polyglot.Value;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Component;
 
+import java.io.IOException;
 import java.nio.file.Path;
 import java.util.Map;
 
 @Component
 public class Invoker {
 
+    private DirigibleJavascriptCodeRunner runner;
+
     private final CamelProcessor processor;
 
     @Autowired
     public Invoker(CamelProcessor processor) {
         this.processor = processor;
     }
 
-    public void invoke(Message camelMessage) {
+    public void invoke(Message camelMessage) throws IOException {
+        resetCodeRunner();
         String resourcePath = (String) camelMessage.getExchange()
                                                    .getProperty("resource");
-        String messageBody = camelMessage.getBody(String.class);
-        try (var runner = new DirigibleJavascriptCodeRunner()) {
-            var module = runner.run(Path.of(resourcePath));
-            runner.runMethod(module, "onMessage", messageBody);
+
+        var module = runner.run(Path.of(resourcePath));
+        var result = runner.runMethod(module, "onMessage", wrapCamelMessage(camelMessage));
+
+        camelMessage.getExchange()
+                    .setMessage(unwrapCamelMessage(result));
+    }
+
+    private IntegrationMessage wrapCamelMessage(Message camelMessage) {
+        return new IntegrationMessage(camelMessage);
+    }
+
+    private Message unwrapCamelMessage(Value value) {
+        if (value.isHostObject()) {
+            IntegrationMessage integrationMessage = value.asHostObject();
+            return integrationMessage.getCamelMessage();
+        } else {
+            throw new IllegalStateException("Unexpected @dirigible/integrations onMessage() return type");
         }
     }
 
+    private void resetCodeRunner() {
```

Comment:
Have you checked whether `Invoker` is a request scoped bean?

Reply:
From what i can see by default, the beans created with @Component are singleton scoped, meaning there will be only one instance per Spring context. There is @RequestScoped annotation that can be used in combination as described here https://stackoverflow.com/questions/74708265/component-with-request-scope-attributes if we need it

Reply:
Request scope is not the only alternative here, it could also be prototype scoped. If you're using the `Invoker` from the Camel Bean component (and not from the regular Class component) you could try and see if Camel requests a new bean every time or it caches it somehow. If using prototype scope results in new `Invoker` instances for every invocation from Camel, then we can tie the lifecycle of the GraalJS context to the Spring Bean lifecycle - https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/context/annotation/Bean.html#destroyMethod()

Reply:
I tested it out by having the Invoker class scoped with @Scope("prototype"). In the debuger i can see that in the **Invoker::invoke()** method **this** is the same object for every http request when in the designer it's singleton and when it's prototype in the designer it's always a different object

Reply:
And have you tried making the `Invoker` implement `AutoCloseable` and see if the close method is invoked as expected?